<a href="https://colab.research.google.com/github/Wipada6430205202/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Mounted at /content/drive


In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย Groupby.apply(list)

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
data.shape

(541909, 8)

###Quiz หาว่ามีกี่ประเทศ

In [ ]:
data['Country'].value_counts().shape

(38,)

## Association Rules

In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=67939128d17aea2d423a78e2a5b41e871f41cc8d56e4a3943df4164ef2232491
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'548768',
 'C544665',
 '553661',
 '556095',
 '559510',
 '556503',
 '568346',
 'C565848',
 '572550',
 '569568',
 '578149',
 '543188',
 '564838',
 '568193',
 '556237',
 '545429',
 '548733',
 '544314',
 '544434',
 '570257',
 '576616',
 '553535',
 '546683',
 '555400',
 '541999',
 'C555253',
 'C541822',
 '543246',
 '564974',
 '559817',
 '559052',
 'C545033',
 '555510',
 '542115',
 '553671',
 '549951',
 '545719',
 '557064',
 '562365',
 '551160',
 '567378',
 '550474',
 '566060',
 '560504',
 '547826',
 '557034',
 '554921',
 'C558255',
 '562534',
 '540026',
 '567670',
 '559489',
 'C555409',
 '566788',
 '572549',
 '564754',
 '558702',
 '544186',
 '553901',
 '570597',
 '545445',
 '538883',
 '562955',
 '537738',
 '541811',
 '556636',
 '567336',
 '543973',
 '576651',
 '575939',
 '558526',
 '552715',
 '544095',
 '552540',
 '576388',
 '552956',
 '544887',
 '563505',
 '561329',
 '571546',
 '573338',
 '566091',
 '543732',
 '566400',
 '553541',
 '545895',
 '551564',
 '574076',
 '554514',
 '555836',
 '5

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

### Quiz 22386 และ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Homework1 หา Frequent Item Set และ Assosiation Rules
-  ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec) [หนามเตย]
-  ของประเทศในทวีป Europe [จจ]
-  ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb) [ใบเตย]
-  ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec) [พิม]
-  ของวันวาเลนไทน์ [ตุนนี่]
-  โดยจัดกลุ่มตามประเทศ+เดือน [พิมจ๊อก]

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# Filter data for the Christmas and New Year period in Europe
start_date = '2014-12-01'
end_date = '2014-12-31'
european_countries = ['Germany', 'France', 'Spain', 'Italy', 'United Kingdom']

filtered_data = data[(data['InvoiceDate'] >= start_date) & (data['InvoiceDate'] <= end_date) & (data['Country'].isin(european_countries))]

# Drop rows with missing or zero Quantity values
filtered_data = filtered_data.dropna(subset=['Quantity'])
filtered_data = filtered_data[filtered_data['Quantity'] > 0]

# Convert Quantity to integer
filtered_data['Quantity'] = filtered_data['Quantity'].astype(int)

# One-hot encode the data for Apriori algorithm
basket = (filtered_data.groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

# Convert quantity values to 0 or 1
basket_sets = basket.applymap(lambda x: 1 if x > 0 else 0)

# Apply Apriori algorithm to find frequent item sets
frequent_itemsets = apriori(basket_sets, min_support=0.01, use_colnames=True)

# Generate association rules
association_rules_result = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

# Filter rules for Europe and Christmas/New Year period
filtered_rules = association_rules_result[association_rules_result['consequents'].astype(str).str.contains('Christmas|New Year')]
filtered_rules = filtered_rules[filtered_rules['antecedents'].astype(str).str.contains('Europe')]

# Display the frequent item sets and association rules
print("Frequent Item Sets:")
print(frequent_itemsets)

print("\nAssociation Rules:")
print(filtered_rules)


##Frequent Item Set และ Assosiation Rules ของประเทศในทวีป Europe


In [2]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [6]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Day  Month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    1     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    1     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   

   Year      Time  
0  2010  08:26:00  
1  2010  08:26:00  
2  2010  08:26:00  
3  2010  08:26:00  
4  2010  0

In [7]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [8]:
EU_country = ['Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [9]:
EU_data_country = data[data['Country']==""]
for i in range(len(EU_country)):
  x = data[data['Country']==EU_country[i]]
  EU_data_country=pd.concat([x,EU_data_country])

In [10]:
EU_data_country

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541194    581493     23204          CHARLOTTE BAG APPLES DESIGN        10   
541195    581493     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR        18   
541196    581493     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
541197    581493     22807         SET OF 6 T-LIGHTS TOADSTOOLS         6   
541198    581493      POST                              POSTAGE         1   

        UnitPrice  CustomerID         Country  Day  Month  Year      Time  
0            2.55     17850.0  United Kingdom    1     12  2010  08:26:00  
1            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
2            2.75     17850.0  United Kingdom    1     12  2010  08:26:00  
3            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
4            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
...           ...         ...             ...  ...    ...   ...       ...  
541194       0.85     12423.0         Belgium    9     12  2011  10:10:00  
541195       0.79     12423.0         Belgium    9     12  2011  10:10:00  
541196       1.25     12423.0         Belgium    9     12  2011  10:10:00  
541197       2.95     12423.0         Belgium    9     12  2011  10:10:00  
541198      15.00     12423.0         Belgium    9     12  2011  10:10:00  

[537957 rows x 11 columns]

In [12]:
EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25712, dtype: object

In [13]:
transactions = EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


In [14]:
transactions

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25712, dtype: object

In [15]:
!pip install apyori

In [67]:
from apyori import apriori


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [68]:
apriori(transactions,min_support = 0.053)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


<generator object apriori at 0x782efcd9a340>

In [69]:
list(apriori(transactions,min_support = 0.053))


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[RelationRecord(items=frozenset({'20725'}), support=0.06191661481020535, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06191661481020535, lift=1.0)]),
 RelationRecord(items=frozenset({'22197'}), support=0.05588830118232732, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22197'}), confidence=0.05588830118232732, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08354075917859365, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08354075917859365, lift=1.0)]),
 RelationRecord(items=frozenset({'22720'}), support=0.05627722464219042, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720'}), confidence=0.05627722464219042, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06568917237087742, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [70]:
frequentItemsets = list(apriori(transactions,min_support = 0.053))
frequentItemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[RelationRecord(items=frozenset({'20725'}), support=0.06191661481020535, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06191661481020535, lift=1.0)]),
 RelationRecord(items=frozenset({'22197'}), support=0.05588830118232732, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22197'}), confidence=0.05588830118232732, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08354075917859365, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08354075917859365, lift=1.0)]),
 RelationRecord(items=frozenset({'22720'}), support=0.05627722464219042, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720'}), confidence=0.05627722464219042, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06568917237087742, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [71]:
len(frequentItemsets)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


8

In [72]:
frequentItemsets[0]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


RelationRecord(items=frozenset({'20725'}), support=0.06191661481020535, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06191661481020535, lift=1.0)])

In [73]:
frequentItemsets[0][0]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


frozenset({'20725'})

In [74]:
dir(frequentItemsets[0][0])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [75]:
frequentItemsets[0][0].__len__()


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


1

In [76]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03235843186060983, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03235843186060983, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674268, lift=8.20174583018949), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.391713747645951, lift=8.20174583018949)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03010267579340386, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03010267579340386, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7406698564593301, lift=17.218899954143126), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.6998191681735986, lift=17.21

In [77]:
set(EU_data_country[(EU_data_country['StockCode']=='85099B')]['Description'])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'JUMBO BAG RED RETROSPOT'}

In [78]:
transactions =EU_data_country.groupby(['Country', 'Year'])['StockCode'].apply(list)
min_sup = 0.4
frequentItemsets= list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


yes
RelationRecord(items=frozenset({'20750', '20685'}), support=0.4146341463414634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20750', '20685'}), confidence=0.4146341463414634, lift=1.0), OrderedStatistic(items_base=frozenset({'20685'}), items_add=frozenset({'20750'}), confidence=0.8095238095238094, lift=1.5086580086580084), OrderedStatistic(items_base=frozenset({'20750'}), items_add=frozenset({'20685'}), confidence=0.7727272727272726, lift=1.5086580086580084)])
yes
RelationRecord(items=frozenset({'21080', '20685'}), support=0.4146341463414634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21080', '20685'}), confidence=0.4146341463414634, lift=1.0), OrderedStatistic(items_base=frozenset({'20685'}), items_add=frozenset({'21080'}), confidence=0.8095238095238094, lift=1.9523809523809523), OrderedStatistic(items_base=frozenset({'21080'}), items_add=frozenset({'20685'}), confidence=1.0, lift=1.9523809523809523)])
y

In [79]:
set(EU_data_country[(EU_data_country['StockCode']=='22807')]['Description'])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'SET OF 6 T-LIGHTS TOADSTOOLS'}

In [80]:
set(EU_data_country[(EU_data_country['StockCode']=='22960')]['Description'])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'JAM MAKING SET WITH JARS'}

In [81]:
EU_data_country[EU_data_country['StockCode']=='22252']

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
9528      537205     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
29874     538827     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        24   
58454     541241     22252  BIRDCAGE DECORATION TEALIGHT HOLDER         4   
64642     541658     22252  BIRDCAGE DECORATION TEALIGHT HOLDER         2   
82479     543235     22252  BIRDCAGE DECORATION TEALIGHT HOLDER         4   
...          ...       ...                                  ...       ...   
270498    560569     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
488728    577850     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
91394     544097     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        96   
96038     544474     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
541196    581493     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   

        UnitPrice  CustomerID         Country  Day  Month  Year      Time  
9528         1.25     13034.0  United Kingdom    5     12  2010  14:52:00  
29874        1.06     14298.0  United Kingdom   14     12  2010  12:59:00  
58454        1.25     13782.0  United Kingdom   16      1  2011  11:08:00  
64642        1.25     15529.0  United Kingdom   20      1  2011  12:16:00  
82479        1.25     14462.0  United Kingdom    4      2  2011  13:30:00  
...           ...         ...             ...  ...    ...   ...       ...  
270498       1.25     12480.0         Germany   19      7  2011  14:04:00  
488728       1.25     14911.0            EIRE   22     11  2011  11:04:00  
91394        1.06     12455.0          Cyprus   16      2  2011  09:56:00  
96038        1.25     12423.0         Belgium   21      2  2011  09:52:00  
541196       1.25     12423.0         Belgium    9     12  2011  10:10:00  

[72 rows x 11 columns]

In [82]:
EU_data_country[EU_data_country['InvoiceNo']=='543235']

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
82452    543235     22948   METAL DECORATION NAUGHTY CHILDREN          1   
82453    543235     21454                 PAINTED PINK RABBIT          1   
82454    543235     22292     HANGING CHICK  YELLOW DECORATION         1   
82455    543235     84799     SPRIG LAVENDER ARTIFICIAL FLOWER        16   
82456    543235    84800L     LARGE WHITE/PINK ROSE ART FLOWER         3   
82457    543235    84800S     SMALL WHITE/PINK ROSE ART FLOWER         6   
82458    543235     84763        ZINC FINISH 15CM PLANTER POTS         4   
82459    543235    47504K        ENGLISH ROSE GARDEN SECATEURS         2   
82460    543235     21745      GAOLERS KEYS DECORATIVE GARDEN          2   
82461    543235     21385      IVORY HANGING DECORATION  HEART         2   
82462    543235    85061W      WHITE JEWELLED HEART DECORATION         1   
82463    543235     84879        ASSORTED COLOUR BIRD ORNAMENT         8   
82464    543235     22292     HANGING CHICK  YELLOW DECORATION         3   
82465    543235     22291       HANGING CHICK CREAM DECORATION         4   
82466    543235     21165        BEWARE OF THE CAT METAL SIGN          1   
82467    543235     22948   METAL DECORATION NAUGHTY CHILDREN          3   
82468    543235     22285       HANGING HEN ON NEST DECORATION         2   
82469    543235     22284                 HEN HOUSE DECORATION         2   
82470    543235     82583                 HOT BATHS METAL SIGN         1   
82471    543235     21166           COOK WITH WINE METAL SIGN          1   
82472    543235     21385      IVORY HANGING DECORATION  HEART         2   
82473    543235     21454                 PAINTED PINK RABBIT          3   
82474    543235     35971       ROSE FOLKART HEART DECORATIONS         4   
82475    543235     22230        JIGSAW TREE WITH WATERING CAN         2   
82476    543235    85061W      WHITE JEWELLED HEART DECORATION         1   
82477    543235    85186A      EASTER BUNNY GARLAND OF FLOWERS         6   
82478    543235     72741                GRAND CHOCOLATECANDLE         2   
82479    543235     22252  BIRDCAGE DECORATION TEALIGHT HOLDER         4   
82480    543235     22173   METAL 4 HOOK HANGER FRENCH CHATEAU         5   
82481    543235     22829            SWEETHEART WIRE WALL TIDY         1   
82482    543235     21340  CLASSIC METAL BIRDCAGE PLANT HOLDER         2   
82483    543235     22789    T-LIGHT HOLDER SWEETHEART HANGING         4   
82484    543235     71459       HANGING JAM JAR T-LIGHT HOLDER        12   
82485    543235     84755  COLOUR GLASS T-LIGHT HOLDER HANGING         8   
82486    543235    82494L          WOODEN FRAME ANTIQUE WHITE          2   
82487    543235     21068      VINTAGE BILLBOARD LOVE/HATE MUG         2   
82488    543235     21871                  SAVE THE PLANET MUG         6   
82489    543235     48138                   DOORMAT UNION FLAG         1   
82490    543235     21355              TOAST ITS - I LOVE YOU          2   
82491    543235     21754             HOME BUILDING BLOCK WORD         1   
82492    543235     22926       IVORY GIANT GARDEN THERMOMETER         4   
82493    543235     21908        CHOCOLATE THIS WAY METAL SIGN         2   
82494    543235     22776          SWEETHEART CAKESTAND 3 TIER         2   
82495    543235     21877                  HOME SWEET HOME MUG         2   
82496    543235     21868                 POTTING SHED TEA MUG         2   
82497    543235     22832            BROCANTE SHELF WITH HOOKS         2   
82498    543235     22281             EASTER TREE YELLOW BIRDS         2   
82499    543235     72741                GRAND CHOCOLATECANDLE         1   
82500    543235    35471D    SET OF 3 BIRD LIGHT PINK FEATHER          1   
82501    543235     22829            SWEETHEART WIRE WALL TIDY         1   
82502    543235     22854          CREAM SWEETHEART EGG HOLDER         2   

       UnitPrice  CustomerID         Country  Day  Mo

## ข้อ 6



In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [4]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [5]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Day  Month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    1     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    1     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   

   Year      Time  
0  2010  08:26:00  
1  2010  08:26:00  
2  2010  08:26:00  
3  2010  08:26:00  
4  2010  0

In [6]:
data_MC=data

In [7]:
data_MC

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

        UnitPrice  CustomerID         Country  Day  Month  Year      Time  
0            2.55     17850.0  United Kingdom    1     12  2010  08:26:00  
1            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
2            2.75     17850.0  United Kingdom    1     12  2010  08:26:00  
3            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
4            3.39     17850.0  United Kingdom    1     12  2010  08:26:00  
...           ...         ...             ...  ...    ...   ...       ...  
541904       0.85     12680.0          France    9     12  2011  12:50:00  
541905       2.10     12680.0          France    9     12  2011  12:50:00  
541906       4.15     12680.0          France    9     12  2011  12:50:00  
541907       4.15     12680.0          France    9     12  2011  12:50:00  
541908       4.95     12680.0          France    9     12  2011  12:50:00  

[541909 rows x 11 columns]

In [8]:
transactions_MC = data_MC.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [9]:
transactions_MC

Country      Month
Australia    1        [22078, 22079, 22080, 22077, 22505, 22516, 225...
             2        [84978, 22960, 47590B, 47566, 22087, 22720, 22...
             3        [22978, 22980, 22993, 22654, 22968, 22505, 225...
             4        [POST, 22232, 22993, 22449, 20726, 20725, 2072...
             5        [22699, 22859, 22730, 84663A, 23206, 48129, 23...
                                            ...                        
Unspecified  6        [23122, 23119, 23108, 23240, 23237, 23239, 232...
             7        [21634, 21212, 22964, 23254, 23256, 84378, 209...
             8        [23236, 21210, 23313, 23318, 22952, 22950, 232...
             9        [22904, 21329, 21992, 20772, 23196, 22543, 225...
             11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
Name: StockCode, Length: 296, dtype: object

In [10]:
!pip install apyori
from apyori import apriori


In [11]:
apriori(transactions_MC,min_support = 0.53)


<generator object apriori at 0x7c47918f5540>

In [12]:
list(apriori(transactions_MC,min_support = 0.53))


[RelationRecord(items=frozenset({'POST'}), support=0.5743243243243243, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST'}), confidence=0.5743243243243243, lift=1.0)])]

In [13]:
frequentItemsets_MC = list(apriori(transactions_MC,min_support = 0.53))
frequentItemsets_MC

[RelationRecord(items=frozenset({'POST'}), support=0.5743243243243243, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST'}), confidence=0.5743243243243243, lift=1.0)])]

In [30]:
min_sup=0.3
frequentItemsets_MC= list(apriori(transactions_MC,min_support =min_sup))
for i in frequentItemsets_MC:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22326', 'POST'}), support=0.32094594594594594, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326', 'POST'}), confidence=0.32094594594594594, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8407079646017699, lift=1.4638209266007287), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5588235294117647, lift=1.463820926600729)])
yes
RelationRecord(items=frozenset({'22423', 'POST'}), support=0.3141891891891892, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423', 'POST'}), confidence=0.3141891891891892, lift=1.0), OrderedStatistic(items_base=frozenset({'22423'}), items_add=frozenset({'POST'}), confidence=0.6838235294117647, lift=1.1906574394463667), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22423'}), confidence=0.5470588235294118, lift=1.1906574394463

In [31]:
data_MC[data_MC['StockCode'] == '85123A'][['Month', 'Country','Description']]


Month         Country                         Description
0          12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
49         12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
66         12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
220        12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
262        12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
...       ...             ...                                 ...
539342     12  United Kingdom  CREAM HANGING HEART T-LIGHT HOLDER
539979     12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
540217     12  United Kingdom  WHITE HANGING HEART T-LIGHT HOLDER
540837     12  United Kingdom  CREAM HANGING HEART T-LIGHT HOLDER
541639     12  United Kingdom  CREAM HANGING HEART T-LIGHT HOLDER

[2313 rows x 3 columns]